v-next/hardhat-keystore/src/internal/ui/user-interactions.ts

```diff
@@ -0,0 +1,120 @@
+import type { UserInterruptionManager } from "@ignored/hardhat-vnext/types/user-interruptions";
+
+import chalk from "chalk";
+
+import { PLUGIN_ID } from "../constants.js";
+
+export class UserInteractions {
+  readonly #userInterruptions: UserInterruptionManager;
+
+  constructor(userInterruptions: UserInterruptionManager) {
+    this.#userInterruptions = userInterruptions;
+  }
+
+  public async setUpPassword(): Promise<string> {
+    const PASSWORD_REGEX = /^(?=.*[A-Z])(?=.*[a-z])(?=.*[\W_]).{8,}$/;
+
+    const setupMsg =
+      "This is the first time you are using the keystore, please set a password.";
+    const passwordRulesMsg =
+      "The password must have at least 8 characters, one uppercase letter, one lowercase letter, and one special character.";
+
+    await this.#displayMessage("\n👷🔐 Hardhat-Keystore 🔐👷\n");
+    await this.#displayMessage(setupMsg);
+    await this.#displayMessage(passwordRulesMsg);
+    await this.#displayMessage("");
+
+    let password: string | undefined;
+
+    while (password === undefined) {
+      password = await this.#requestSecretInput(`Enter your password: `);
+
+      if (!PASSWORD_REGEX.test(password)) {
```

Comment:
I think this is application logic, and shouldn't be hidden here in this UI-level module. I completely missed it until @ChristopherDedominici corrected me that the other regex validates the key, not the password.

Reply:
We have talked it over. This code is not run in this version, but we didn't want to lose it. For now it is used only from a test. So we could delete it and bring it back later.

However, we are suggesting the next steps of:

1. Create a util helper function `isValidPassword` that (for the moment) returns a boolean
2. Inject in the validation function as a parameter to `setupPassword`